returns/contrib/sphinx_autodoc_typehints.py

```diff
@@ -0,0 +1,459 @@
+
+    add(comment[start_arg_at: at + 1])
+    return result
+
+
+def process_docstring(app, what, name, obj, options, lines):
+    original_obj = obj
+    if isinstance(obj, property):
+        obj = obj.fget
+
+    if callable(obj):
+        if inspect.isclass(obj):
+            obj = getattr(obj, '__init__')
+
+        obj = inspect.unwrap(obj)
+        type_hints = get_all_type_hints(obj, name)
+
+        for argname, annotation in type_hints.items():
+            if argname == 'return':
+                continue  # this is handled separately later
+            if argname.endswith('_'):
+                argname = '{}\\_'.format(argname[:-1])
+
+            formatted_annotation = format_annotation(
+                annotation, fully_qualified=app.config.typehints_fully_qualified)
+
+            searchfor = [':{} {}:'.format(field, argname)
+                         for field in ('param', 'parameter', 'arg', 'argument')]
+            insert_index = None
+
+            for i, line in enumerate(lines):
+                if any(line.startswith(search_string) for search_string in searchfor):
+                    insert_index = i
+                    break
+
+            if insert_index is None and app.config.always_document_param_types:
+                lines.append(':param {}:'.format(argname))
+                insert_index = len(lines)
+
+            if insert_index is not None:
+                lines.insert(
+                    insert_index,
+                    ':type {}: {}'.format(argname, formatted_annotation)
+                )
+
+        if 'return' in type_hints and not inspect.isclass(original_obj):
+            # This avoids adding a return type for data class __init__ methods
+            if what == 'method' and name.endswith('.__init__'):
+                return
+
+            formatted_annotation = format_annotation(
+                type_hints['return'], fully_qualified=app.config.typehints_fully_qualified)
+
+            insert_index = len(lines)
+            for i, line in enumerate(lines):
+                if line.startswith(':rtype:'):
+                    insert_index = None
+                    break
+                elif line.startswith(':return:') or line.startswith(':returns:'):
+                    insert_index = i
+
+            if insert_index is not None and app.config.typehints_document_rtype:
+                if insert_index == len(lines):
+                    # Ensure that :rtype: doesn't get joined with a paragraph of text, which
+                    # prevents it being interpreted.
+                    lines.append('')
+                    insert_index += 1
+
+                lines.insert(insert_index, ':rtype: {}'.format(formatted_annotation))
+
+        # Introduce a blank line if the last line has content
+        if lines[-1]:
+            lines.append('')
+
+
+def builder_ready(app):
+    if app.config.set_type_checking_flag:
+        typing.TYPE_CHECKING = True
+
+
+def setup(app):
```

Comment:
Let's better turn off this `fail_on_warnings` thing! This looks scary!
I don't want to vendor this much code 😞

Reply:
Good news, I realized we can do this (fix the warnings) inside our `conf.py` file!! I'll code asap to update this PR

Reply:
I've created a very minimal plugin to deal with it